From 424879fe9dc51c4775a9df48240cffb84cd3335f Mon Sep 17 00:00:00 2001
From: Julien STEPHAN <jstephan@baylibre.com>
Date: Thu, 25 Feb 2021 14:29:07 +0000
Subject: [PATCH] add external gpu delegate

Add an external gpu delegate to wrap the gpu delegate, so the gpu
delegate can be used with python bindings

Signed-off-by: Julien STEPHAN <jstephan@baylibre.com>
---
 .../utils/gpu_external_delegate/BUILD         | 27 ++++++++++++++
 .../external_delegate_adaptor.cc              | 36 +++++++++++++++++++
 2 files changed, 63 insertions(+)
 create mode 100644 tensorflow/lite/delegates/utils/gpu_external_delegate/BUILD
 create mode 100644 tensorflow/lite/delegates/utils/gpu_external_delegate/external_delegate_adaptor.cc

diff --git a/tensorflow/lite/delegates/utils/gpu_external_delegate/BUILD b/tensorflow/lite/delegates/utils/gpu_external_delegate/BUILD
new file mode 100644
index 00000000000..4c9c0cdf817
--- /dev/null
+++ b/tensorflow/lite/delegates/utils/gpu_external_delegate/BUILD
@@ -0,0 +1,27 @@
+load("//tensorflow/lite:build_def.bzl", "tflite_copts")
+load("//tensorflow/lite/tools/evaluation/tasks:build_def.bzl", "task_linkopts")
+
+package(
+    default_visibility = [
+        "//visibility:public",
+    ],
+    licenses = ["notice"],  # Apache 2.0
+)
+
+cc_binary(
+    name = "gpu_external_delegate.so",
+    srcs = [
+        "external_delegate_adaptor.cc",
+    ],
+    linkopts = [
+        "-Wl,-soname=gpu_external_delegate.so",
+    ],
+    linkshared = 1,
+    linkstatic = 1,
+    deps = [
+        "//tensorflow/lite/delegates/gpu:delegate",
+        "//tensorflow/lite/c:common",
+        "//tensorflow/lite/tools:command_line_flags",
+        "//tensorflow/lite/tools:logging",
+    ],
+)
diff --git a/tensorflow/lite/delegates/utils/gpu_external_delegate/external_delegate_adaptor.cc b/tensorflow/lite/delegates/utils/gpu_external_delegate/external_delegate_adaptor.cc
new file mode 100644
index 00000000000..8900a15b2ca
--- /dev/null
+++ b/tensorflow/lite/delegates/utils/gpu_external_delegate/external_delegate_adaptor.cc
@@ -0,0 +1,36 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/lite/delegates/gpu/delegate.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Defines two symbols that need to be exported to use the TFLite external
+// delegate. See tensorflow/lite/delegates/external for details.
+TFL_CAPI_EXPORT TfLiteDelegate* tflite_plugin_create_delegate(
+    char** options_keys, char** options_values, size_t num_options,
+    void (*report_error)(const char*)) {
+  return TfLiteGpuDelegateV2Create(nullptr);
+}
+
+TFL_CAPI_EXPORT void tflite_plugin_destroy_delegate(TfLiteDelegate* delegate) {
+	TfLiteGpuDelegateV2Delete(delegate);
+}
+
+#ifdef __cplusplus
+}
+#endif  // __cplusplus
-- 
2.17.1

